### ISSUE TITLE
How to Freeze Detection Head Layers in YOLOv8m-segment and Train Only Segmentation Head?

### ISSUE THREAD

**Issue by Wang-taoshuo** [ultralytics/ultralytics#19826]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi all,

I'm working with yolov8m-seg.pt and want to freeze the detection head layers (bounding box/class prediction) while training only the segmentation head (mask prediction). The goal is to fine-tune the segmentation capability without updating the detection part. Has anyone done this before?

I’m thinking of freezing layers by setting requires_grad = False for detection-related params, but I’m unsure how to precisely identify them in the head (e.g., model.22). Here’s my tentative code—can someone confirm if this approach works or suggest a better way?

### Additional

`from ultralytics import YOLO

# Load model
model = YOLO("yolov8m-seg.pt")

# Freeze detection head layers (guessing these are related to 'detect')
for name, param in model.model.named_parameters():
    if "detect" in name.lower():  # Is this the right way to target detection head?
        param.requires_grad = False

# Train only segmentation head
model.train(data="path/to/data.yaml", epochs=50, imgsz=640)`
Questions:

Does detect correctly target the detection head, or should I use a different identifier (e.g., specific layer indices)?
Will this setup ensure the segmentation head (e.g., mask coefficients/Proto) still trains properly?
Any pitfalls to watch out for?
Thanks for any insights!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Wang-taoshuo, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance, where you can explore detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions and best practices for YOLO use cases, including segmentation tasks, are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

For custom training or advanced configuration ❓ Questions like yours, please ensure you provide as much detail as possible, including the following:
- Dataset structure and sample images
- Training logs and environment details
- Any modifications made to the model or training code

Your approach to freezing layers seems reasonable, but we suggest carefully examining the model architecture to ensure you're targeting the correct layers. This can be done using `model.model` to inspect the components and their names. Additionally, you can review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to optimize training.

To further enhance your YOLO experience, consider joining the Ultralytics community:
- For real-time chat and support: [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions: [Discourse](https://community.ultralytics.com/)
- To share knowledge and tips: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are using the latest ultralytics package by upgrading it along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following environments with pre-installed dependencies ([CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to guide you in the right direction. An Ultralytics engineer will review and assist you further soon. Thank you for supporting Ultralytics! 😊

**Comment by Wang-taoshuo** [comment]
Hello, could you please review my code to confirm that it correctly replaces the backbone and neck components in the yolov8-seg.pt model? Additionally, I have one more question: I’m not sure how to freeze the object detection head.
"import os
import shutil
import yaml
import torch
from ultralytics import YOLO
import matplotlib.pyplot as plt
import seaborn as sns

# ------------------ 訓練與融合模組 ------------------

def fuse_detection_weights_to_segmentation(
    detection_model_path,      # 物件偵測模型路徑
    segmentation_model_path,   # 分割模型路徑（官方預訓練，80類）
    output_model_path,         # 融合後模型保存路徑
    data_yaml_path,            # 分割資料集 YAML 文件路徑
    freeze_backbone=True,      # 是否凍結 backbone 與 neck
    initial_epochs=50,         # 第一階段訓練 (僅訓練分割 head) 的 epoch 數
    fine_tuning_epochs=50      # 第二階段訓練 (僅微調分割 head) 的 epoch 數
):
    """
    融合流程：
      1. 加載物件偵測模型與分割模型
      2. 取出檢測模型的權重，僅替換共享層（backbone 與 neck），保留分割模型 head
      3. 保存融合後模型
      4. 第一階段訓練：凍結 backbone 與 neck，只訓練分割 head
      5. 第二階段訓練：以較低學習率微調分割 head，保持 backbone 與 neck 冻結
    """
    print("📋 開始模型融合流程...")
    # 檢查文件是否存在
    for path, desc in [(detection_model_path, "物件偵測模型"),
                       (segmentation_model_path, "分割模型"),
                       (data_yaml_path, "資料集配置文件")]:
        assert os.path.exists(path), f"{desc}不存在: {path}"
    
    os.makedirs(os.path.dirname(output_model_path), exist_ok=True)
    
    # 1. 加載物件偵測模型
    print(f"🔄 加載物件偵測模型: {detection_model_path}")
    detection_model = YOLO(detection_model_path)
    
    # 2. 加載分割模型
    print(f"🔄 加載分割模型: {segmentation_model_path}")
    segmentation_model = YOLO(segmentation_model_path)
    
    # 3. 取得狀態字典
    detection_state_dict = detection_model.model.state_dict()
    segmentation_state_dict = segmentation_model.model.state_dict()
    
    # 4. 創建共享層映射表（僅替換 backbone 與 neck）
    shared_layers = {}
    print("🔍 分析共享層結構...")
    for name in detection_state_dict.keys():
        # 排除 head 部分（假設 head 部分以 "model.22" 開頭）
        if name in segmentation_state_dict and not name.startswith('model.22'):
            shared_layers[name] = detection_state_dict[name]
    print(f"✅ 找到 {len(shared_layers)} 個共享層可融合")
    
    # 5. 更新分割模型的共享層，保留 head 部分不變
    updated_weights = {}
    for name, param in segmentation_state_dict.items():
        if name.startswith('model.22'):
            # 保留分割 head 權重（不進行替換）
            updated_weights[name] = param
        elif name in shared_layers:
            updated_weights[name] = shared_layers[name]
        else:
            updated_weights[name] = param
    
    segmentation_model.model.load_state_dict(updated_weights)
    print("✅ 權重融合完成")
    
    # 6. 保存融合後的模型暫存檔
    temp_fusion_path = os.path.join(os.path.dirname(output_model_path), 'temp_fusion.pt')
    segmentation_model.save(temp_fusion_path)
    print(f"💾 融合模型暫存至: {temp_fusion_path}")
    
    # 7. 第一階段訓練 - 僅訓練分割 head (保持 backbone 與 neck 凍結)
    print("\n🚀 開始第一階段訓練 (凍結 backbone 與 neck, 僅訓練分割 head)...")
    fused_model = YOLO(temp_fusion_path)
    # 設置凍結層，這裡凍結所有 backbone 與 neck 層 (假設 head 層在 model.22 開頭)
    freeze_layers = []
    if freeze_backbone:
        model_modules = list(fused_model.model.modules())
        for i, module in enumerate(model_modules):
            if any(x in str(type(module)) for x in ['Backbone', 'Neck', 'C2f', 'Conv']):
                freeze_layers.append(i)
    freeze_string = ','.join(map(str, freeze_layers[:10])) + '...' if len(freeze_layers) > 10 else ','.join(map(str, freeze_layers))
    print(f"🧊 凍結層: {freeze_string}")
    
    phase1_results = fused_model.train(
        data=data_yaml_path,
        epochs=initial_epochs,
        imgsz=640,
        batch=16,
        patience=15,
        freeze=freeze_layers,  # 保持凍結
        project=os.path.dirname(output_model_path),
        name='phase1_auto_training',
        exist_ok=True,
        verbose=True
    )
    
    # 8. 第二階段訓練 - 繼續僅訓練分割 head，使用較低學習率微調
    if fine_tuning_epochs > 0:
        print("\n🚀 開始第二階段訓練 (微調分割 head，保持 backbone 與 neck 凍結)...")
        phase1_best_model_path = os.path.join(os.path.dirname(output_model_path), 'phase1_auto_training', 'weights', 'best.pt')
        phase2_model = YOLO(phase1_best_model_path)
        
        phase2_results = phase2_model.train(
            data=data_yaml_path,
            epochs=fine_tuning_epochs,
            imgsz=640,
            batch=16,  # 較小的 batch size 以利精細調整
            patience=50,
            freeze=freeze_layers,  # 繼續凍結
            lr0=0.001,            # 降低初始學習率
            lrf=0.1,
            weight_decay=0.0005,
            optimizer='auto',
            #optimizer='AdamW',
            project=os.path.dirname(output_model_path),
            name='phase2_auto_training',
            exist_ok=True,
            verbose=True
        )
        
        final_best_model_path = os.path.join(os.path.dirname(output_model_path), 'phase2_auto_training', 'weights', 'best.pt')
        shutil.copy(final_best_model_path, output_model_path)
    else:
        phase1_best_model_path = os.path.join(os.path.dirname(output_model_path), 'phase1_auto_training', 'weights', 'best.pt')
        shutil.copy(phase1_best_model_path, output_model_path)
    
    # 9. 清理暫存文件
    if os.path.exists(temp_fusion_path):
        os.remove(temp_fusion_path)
    
    print(f"\n✅ 模型融合與訓練流程完成!")
    print(f"📊 最終模型已保存至: {output_model_path}")
    return output_model_path

# ------------------ 預測模組 ------------------
def visualize_segmentation_performance(model_path, test_images_dir, output_dir):
    """
    評估融合後分割模型的效能並可視化結果
    """
    print("\n📊 開始評估分割模型效能...")
    os.makedirs(output_dir, exist_ok=True)
    
    model = YOLO(model_path)
    
    results = model.predict(
        test_images_dir, 
        save=True, 
        save_txt=True, 
        project=output_dir, 
        name='test_results'
    )
    print("Compare heads results:", results)
    metrics = model.val(
        data=os.path.join(os.path.dirname(test_images_dir), 'data.yaml')
    )
    
    print(f"✅ 模型評估完成，結果已保存至: {os.path.join(output_dir, 'test_results')}")
    return metrics

# ------------------ 主流程 ------------------

def main():
    # 讀取 YAML 配置檔
    config_path = r"D:\20250320\20250327meeting\code\config.yaml"  # 請確保此文件存在且內容正確
    with open(config_path, "r", encoding="utf-8") as f:
        config = yaml.safe_load(f)
    
    # 提取訓練相關參數
    training_cfg = config["training"]
    detection_model_path = training_cfg["detection_model_path"]
    segmentation_model_path = training_cfg["segmentation_model_path"]
    output_model_path = training_cfg["output_model_path"]
    data_yaml_path = training_cfg["data_yaml_path"]
    freeze_backbone = training_cfg.get("freeze_backbone", True)
    initial_epochs = training_cfg.get("initial_epochs", 50)
    fine_tuning_epochs = training_cfg.get("fine_tuning_epochs", 50)
    
    # 執行模型融合與訓練
    fused_model_path = fuse_detection_weights_to_segmentation(
        detection_model_path=detection_model_path,
        segmentation_model_path=segmentation_model_path,
        output_model_path=output_model_path,
        data_yaml_path=data_yaml_path,
        freeze_backbone=freeze_backbone,
        initial_epochs=initial_epochs,
        fine_tuning_epochs=fine_tuning_epochs
    )
    
    # 提取預測相關參數
    prediction_cfg = config["prediction"]
    test_images_dir = prediction_cfg["test_images_dir"]
    visualization_dir = prediction_cfg["visualization_dir"]

    # 執行模型評估與可視化
    metrics = visualize_segmentation_performance(
        model_path=fused_model_path,
        test_images_dir=test_images_dir,
        output_dir=visualization_dir
    )
    print("\n📋 融合模型評估指標:")
    print(f"mAP50: {metrics.box.map50:.4f}")
    print(f"mAP50-95: {metrics.box.map:.4f}")
    print(f"Segmentation mAP50: {metrics.seg.map50:.4f}")
    print(f"Segmentation mAP50-95: {metrics.seg.map:.4f}")

if __name__ == "__main__":
    main()
"

**Comment by Y-T-G** [comment]
You can try freezing the convolutions for the detection branch.

```python
from ultralytics import YOLO, ASSETS
model = YOLO("yolo11n-seg.pt")
freeze = [f"{k}" for k in range(23)]
freeze += ["23.cv2", "23.cv3"]
results = model.train(data="coco128-seg.yaml", freeze=freeze, epochs=5, exist_ok=True)
```

You would also need to freeze the BatchNorm layers manually.

**Comment by Jordan-Pierce** [comment]
Thoughts on having freeze take an optional float value 0-1 representing the percentage of layers to free within the encoder? Targeting a specific number of layers without knowing prior is kind of annoying, but if I know I want to freeze most of the layers in the encoder, I can just set it to 0.95? Additionally, having an option to freeze the encoder (bool) would be helpful.

**Comment by glenn-jocher** [comment]
@Jordan-Pierce thanks for your suggestion! While the current `freeze` parameter requires explicit layer indices, you can achieve percentage-based freezing by first inspecting your model's architecture via `model.model.named_parameters()` to identify layer indices, then passing the desired subset to `freeze=[...]`. For encoder-specific freezing, we recommend using the existing layer index approach as encoder implementations vary between YOLO versions. 

If you share which specific YOLO version/model you're using, we can help identify the relevant encoder layers to freeze. For feature requests, please feel free to open a GitHub Issue describing your use case at [Ultralytics YOLO Issues](https://github.com/ultralytics/ultralytics/issues/new/choose).

**Comment by Y-T-G** [comment]
> Thoughts on having freeze take an optional float value 0-1 representing the percentage of layers to free within the encoder? Targeting a specific number of layers without knowing prior is kind of annoying, but if I know I want to freeze most of the layers in the encoder, I can just set it to 0.95? Additionally, having an option to freeze the encoder (bool) would be helpful.

I think you can achieve that easily with little code:
```
num_freeze = len(model.model) * 0.95
results = model.train(..., freeze=freeze)
```

**Comment by Jordan-Pierce** [comment]
> > Thoughts on having freeze take an optional float value 0-1 representing the percentage of layers to free within the encoder? Targeting a specific number of layers without knowing prior is kind of annoying, but if I know I want to freeze most of the layers in the encoder, I can just set it to 0.95? Additionally, having an option to freeze the encoder (bool) would be helpful.
> 
> I think you can achieve that easily with little code:
> ```
> num_freeze = int(len(model.model.model) * 0.95)
> results = model.train(..., freeze=num_freeze)
> ```

It's what I ended up doing.

**Comment by glenn-jocher** [comment]
Great to hear you found a working solution! Your approach using `len(model.model.model)` to calculate percentage-based freezing is indeed the recommended way with current functionality. For formalizing this as a feature, feel free to open a GitHub Issue describing your use case at [Ultralytics YOLO Issues](https://github.com/ultralytics/ultralytics/issues/new/choose) - we welcome community input on improving training flexibility. Let us know if you need any clarification on layer identification for your specific model! 🚀

**Comment by Wang-taoshuo** [comment]
Hi @glenn-jocher ,

I'm integrating a pre-trained YOLOv8 detection model into a segmentation model by transferring the backbone, neck, and detection head weights while keeping the segmentation head (model.22.proto and model.22.cv4) trainable. I've followed your recommendations to freeze the detection branch during training.

However, I've noticed that the weights for the very first layers, specifically model.0.conv.bias and model.0.conv.weight, still show a significant maximum difference (around 15.2 and 10.2 respectively) even after freezing. I expected that these early backbone layers would remain unchanged when properly frozen.

Could you please advise if there are any additional considerations or modifications needed to ensure that the backbone layers are truly frozen during segmentation head training? Any insights on why these initial layers might be exhibiting such marked differences would be greatly appreciated.

Thank you for your time and help.

Best regards,

`def freeze_detection_branch(model):
    for name, param in model.named_parameters():
        # Keep parameters of the segmentation head (starting with "model.22.proto" or "model.22.cv4") trainable
        if name.startswith("model.22.proto") or name.startswith("model.22.cv4"):
            param.requires_grad = True
        else:
            param.requires_grad = False
    # Set all BatchNorm layers to eval mode to freeze their running statistics
    for module in model.modules():
        if isinstance(module, torch.nn.BatchNorm2d):
            module.eval()
    print("[Info] Detection branch has been frozen. Only the segmentation head will be trained.")
    return model
`
